mmhuman3d/core/visualization/renderer/torch3d_renderer/uv_renderer.py

```diff
@@ -0,0 +1,517 @@
+            param_dict = dict(np.load(uv_param_path))
+
+            verts_uv = torch.Tensor(param_dict['verts_uv'])
+            verts_u, verts_v = torch.unbind(verts_uv, -1)
+            verts_v_ = 1 - verts_u.unsqueeze(-1)
+            verts_u_ = verts_v.unsqueeze(-1)
+            self.verts_uv = torch.cat([verts_u_, verts_v_], -1).to(self.device)
+            self.faces_uv = torch.LongTensor(param_dict['faces_uv']).to(
+                self.device)
+
+            self.NUM_VT = self.verts_uv.shape[0]
+
+            self.faces_tensor = torch.LongTensor(param_dict['faces'].astype(
+                np.int64)).to(self.device)
+            self.num_faces = self.faces_uv.shape[0]
+        elif obj_path is not None:
+            check_path_suffix(obj_path, allowed_suffix=['obj'])
+            mesh_template = load_objs_as_meshes([obj_path])
+            self.faces_uv = mesh_template.textures.faces_uvs_padded()[0].to(
+                self.device)
+            self.verts_uv = mesh_template.textures.verts_uvs_padded()[0].to(
+                self.device)
+            self.NUM_VT = self.verts_uv.shape[0]
+            self.faces_tensor = mesh_template.faces_padded()[0].to(self.device)
+            self.num_faces = self.faces_uv.shape[0]
+        self.update_fragments()
+        self.update_face_uv_pixel()
+
+        self = self.to(self.device)
+
+    def to(self, device):
+        if isinstance(device, str):
+            device = torch.device(device)
+        self.device = device
+        for k in dir(self):
+            if isinstance(getattr(self, k), (torch.Tensor)):
+                setattr(self, k, getattr(self, k).to(device))
+        return self
+
+    def update_fragments(self):
+        """Update pix_to_face, bary_coords."""
+        rasterizer = MeshRasterizer(
+            cameras=FoVOrthographicCameras(
+                min_x=1, max_x=0, max_y=1, min_y=0, device=self.device),
+            raster_settings=RasterizationSettings(
+                blur_radius=0,
+                image_size=self.resolution,
+                faces_per_pixel=1,
+                perspective_correct=False,
+            )).to(self.device)
+        verts_uv = torch.cat([
+            self.verts_uv[None],
+            torch.ones(1, self.NUM_VT, 1).to(self.device)
+        ], -1)
+
+        fragments = rasterizer(
+            Meshes(verts=verts_uv, faces=self.faces_uv[None]))
+        self.pix_to_face = fragments.pix_to_face[0, ..., 0]
+        self.bary_coords = fragments.bary_coords[0, ..., 0, :]
+        self.mask = (self.pix_to_face >= 0).long()
+
+    def update_face_uv_pixel(self):
+        """Move the pixels lie on the edges inside the mask, then refine the
+        rest points by searching the nearest pixel in the faces it should be
+        in."""
+        H, W = self.resolution
+        device = self.device
+        cameras = FoVOrthographicCameras(
+            min_x=1, max_x=0, max_y=1, min_y=0, device=self.device)
```

Comment:
min_x > max_x

Reply:
This is right because it need to be flipped